owr/owr.c

```diff
@@ -241,59 +237,78 @@ void owr_init()
+    g_async_queue_push(msg_queue, "ready");
+    return FALSE;
 }
 
+static gpointer owr_run_thread_func(GAsyncQueue *msg_queue)
+{
+    GSource *idle_source;
+
+    g_return_val_if_fail(msg_queue, NULL);
+    g_return_val_if_fail(owr_main_context, NULL);
+
+    idle_source = g_idle_source_new();
+    g_source_set_callback(idle_source, (GSourceFunc) owr_running_callback, msg_queue, NULL);
+    g_source_set_priority(idle_source, G_PRIORITY_DEFAULT);
+    g_source_attach(idle_source, owr_main_context);
+    msg_queue = NULL;
+
+    owr_run();
+    return NULL;
+}
 
 /**
- * owr_init_with_main_context:
- * @main_context: a #GMainContext to be used in OpenWebRTC.
+ * owr_run:
  *
- * Initializes the OpenWebRTC library with the given #GMainContext. When this
- * function is used the application is responsible for iterating the
- * #GMainContext. Either This function or owr_init() must be called before
- * doing anything else.
+ * Runs the OpenWebRTC main-loop inside the current thread.
  */
-void owr_init_with_main_context(GMainContext *main_context)
+void owr_run(void)
 {
-    g_return_if_fail(!owr_initialized);
-    g_return_if_fail(main_context);
+    g_return_if_fail(owr_main_context);
 
-    owr_main_context = main_context;
-    owr_init();
+    owr_main_loop = g_main_loop_new(owr_main_context, FALSE);
+    g_main_loop_run(owr_main_loop);
 }
 
-#if defined(__ANDROID__)
-static gboolean keep_alive_function(gpointer data) {
-    /* I keep the mainloop thread warm */
-    OWR_UNUSED(data);
-    return TRUE;
+/**
+ * owr_run_in_background:
+ *
+ * Creates a new thread and runs the OpenWebRTC main-loop inside that thread.
+ * This function does not return until the thread has started and the mainloop is running.
+ */
+void owr_run_in_background(void)
+{
+    GAsyncQueue *msg_queue = g_async_queue_new();
+
+    g_return_if_fail(owr_main_context);
+
+    owr_main_thread = g_thread_new("owr_main_loop", (GThreadFunc) owr_run_thread_func, msg_queue);
+    g_async_queue_pop(msg_queue);
+    g_async_queue_unref(msg_queue);
 }
-#endif
 
-static gpointer owr_run(gpointer data)
+/**
+ * owr_quit:
+ *
+ * Quits the OpenWebRTC main-loop, and stops the background thread if owr_run_in_background was used.
+ */
+void owr_quit(void)
 {
-#if defined(__ANDROID__)
-    /* Fix for the mainloop thread not always being awoken when adding a source
-        https://bugzilla.gnome.org/show_bug.cgi?id=745965 */
-    GSource *keep_alive_timer = g_timeout_source_new(1000);
-    g_source_set_callback(keep_alive_timer, keep_alive_function, NULL, NULL);
-    g_source_attach(keep_alive_timer, owr_main_context);
-#endif
-    g_return_val_if_fail(!data, NULL);
-    g_main_context_push_thread_default(owr_main_context);
-    g_main_loop_run(owr_main_loop);
-    g_main_context_pop_thread_default(owr_main_context);
-    return NULL;
+    g_return_if_fail(owr_main_loop);
+
+    g_main_loop_quit(owr_main_loop);
+    g_main_loop_unref(owr_main_loop);
+    owr_main_loop = NULL;
```

Comment:
And don't unref/set to NULL here. It's safer to do it all after the main loop is done running for real